### PR TITLE
Create a value object for a component record

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes

Create a new `JComponentRecord` value object representing the object loaded by `JComponentHelper::load()` for a component.  Shift the logic for converting the record's parameters to a `Registry` object to the point where the parameters are first accessed (this removes a needless conversion if calling `JComponentHelper::isEnabled()` and that is the only interaction with a component's record).

The value object extends `JObject` as a bridge to help with the transition.  At 4.0, this class extension should be dropped as well as the magic getter/setter methods.

### Testing Instructions

The CMS still functions as intended.  Component parameters are lazy loaded when needed.

### Documentation Changes Required

Document the new class and scheduled deprecations.

### CODE REVIEW

File: libraries/cms/component/record.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+defined('JPATH_PLATFORM') or die;
+
+use Joomla\Registry\Registry;
+
+/**
+ * Object representing a component extension record
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class JComponentRecord extends JObject
```
Comment (wilsonge) [inline libraries/cms/component/record.php]: I don't understand why this needs to extend JObject?

Comment (mbabker) [inline libraries/cms/component/record.php]: Erred heavily on the side of caution since right now it's a stdClass object.  Much of the same reasoning why the new `JMenuItem` does the same.  The intent is in 4.0 to remove this.

If it's not needed I can drop it.

Comment (wilsonge) [inline libraries/cms/component/record.php]: OK - I can buy it. In that case can we please add a `@note` in to reflect we are going to drop `JObject` property in J4 then please

Comment (dgt41) [inline libraries/cms/component/record.php]: Can we do the same for XTD_buttons?

Comment (mbabker) [inline libraries/cms/component/record.php]: @dgt41 Make your own PR 😛

**Comment by mbabker** [comment]
FWIW I think we should make more liberal use of value objects versus untyped `stdClass` objects as stated in https://github.com/joomla/joomla-cms/issues/13922 and not just for the lazy loading behaviors that have conveniently come here and the menu item's object.  It also helps to make it clear exactly what properties you're getting on an object versus having to reverse engineer the code to find what is actually being loaded into memory and made available, so it makes it a little easier to jump into the code too.  Developer Experience improvements FTW.

File: libraries/cms/component/record.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	 *
+	 * @var    integer
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public $enabled;
+
+	/**
+	 * Class constructor
+	 *
+	 * @param   array  $data  The menu item data to load
```
Comment (joomdonation) [inline libraries/cms/component/record.php]: Look like copy/paste comment from somewhere. Should be The component record data to load or something like that

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 53a260f5e5c1f4df5f527212be3354e13c2abd45<br /><br />1. Access to some pages in frontend, backend, creating new articles and all still working as expected

2. Write some custom code to check and confirm that params is lazy load<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14044">issues.joomla.org/tracker/joomla-cms/14044</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on ff7e2c02bee431c61fb29916bb93c33cea9fc3fe<br /><br />Click on different Menues in Frontend, append Menu and Modules.

Don't tested lazy load cause don't know how.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14044">issues.joomla.org/tracker/joomla-cms/14044</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC as there are 2 successfully Tests?

**Comment by jeckodevelopment** [comment]
I have tested this item :white_check_mark: successfully on ff7e2c02bee431c61fb29916bb93c33cea9fc3fe<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14044">issues.joomla.org/tracker/joomla-cms/14044</a>.</sub>

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14044">issues.joomla.org/tracker/joomla-cms/14044</a>.</sub>

**Comment by rdeutz** [comment]
Agreed on the concept, but because JObject is deprecated in 4.0, why we are doing it? Or will you remove the extension of JObject in 4.0 and keep the value object?

**Comment by mbabker** [comment]
The extension of JObject is just to be extremely safe about it (we're converting from a `stdClass` to the value object).  The extension goes away in 4.0 and the object stays.

**Comment by rdeutz** [comment]
ok, so I got the idea
